### PR TITLE
monitor tabs tests

### PR DESCRIPTION
**Pull request by will-weiss**

### CODE REVIEW

File: src/tests/integration/monitor-tabs.test.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+  describe('chrome.windows.onCreated', () => {
+    it("Adds a newly created window to store's state", () => {
+      const [callback] = mocks.chrome.windows.onCreated.addListener.firstCall.args
+
+      const win = { id: 4, focused: true }
+      callback(win)
+      const state = mocks.getState()
+
+      if (state.mostRecentAction.type !== 'chrome.windows.onCreated') {
```
Comment (will-weiss) [inline src/tests/integration/monitor-tabs.test.ts:162]: There are 2 cases we want to test. One where `focused: true` and one where `focused: false`. In the first case the focusedWindowId should now be updated, but not in the second case. We assume that the `mostRecentAction` works, so we don't need to test that.

File: src/tests/integration/monitor-tabs.test.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    })
+  })
+
+  describe('chrome.tabs.onRemoved', () => {
+    it("removes a tab from the store's state", () => {
+      const [callbackOnRemove] = mocks.chrome.tabs.onRemoved.addListener.firstCall.args
+
+      const tabId = 11
+      const removeInfo = {
+        id: 11,
```
Comment (will-weiss) [inline src/tests/integration/monitor-tabs.test.ts:35]: Generally looks good, although let's reduce the amount of code we need by defining only the variables/properties that are essential to the test. When we look we see `removeInfo` is never even used, so we can simplify like so.

```typescript
const toRemoveId = 11
expect(mocks.getState().tabs.get(toRemoveId)).to.not.equal(undefined)
const [callback] = mocks.chrome.tabs.onRemoved.addListener.firstCall.args
callback(toRemoveId)
expect(mocks.getState().tabs.get(toRemoveId)).to.equal(undefined)
```

File: src/tests/integration/monitor-tabs.test.ts [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+  describe('chrome.tabs.onUpdated', () => {
+    it("updates a tab change to store's state", () => {
+      const [callbackOnUpdated] = mocks.chrome.tabs.onUpdated.addListener.firstCall.args
+
+      const tabId = 16
+      const changeInfo = {
+        id: 16,
+        windowId: 3,
+        active: false,
```
Comment (will-weiss) [inline src/tests/integration/monitor-tabs.test.ts:57]: We only care about url changes, so let's simplify this.

```typescript
const changeInfo = { id: 16, url: 'https://changeinfo.com/abc' }
callback(changeInfo.id, changeInfo)
```

File: src/tests/integration/monitor-tabs.test.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+  describe('chrome.tabs.onAttached', () => {
+    it("Attaches a tab to another window and updates the store's state", () => {
+      const [callbackOnAttached] = mocks.chrome.tabs.onAttached.addListener.firstCall.args
+
+      const tabId = 15
+      const attachInfo = {
+        id: 15,
+        windowId: 2,
+        newWindowId: 3,
+        active: false,
```
Comment (will-weiss) [inline src/tests/integration/monitor-tabs.test.ts:82]: remove active and url properties

File: src/tests/integration/monitor-tabs.test.ts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        url: 'https://slam.com/abc',
+      }
+
+      callbackOnAttached(tabId, attachInfo)
```
Comment (will-weiss) [inline src/tests/integration/monitor-tabs.test.ts:86]: `callback(attachInfo.id, attachInfo)`

File: src/tests/integration/monitor-tabs.test.ts [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+      const state = mocks.getState()
+      const tab = state.tabs.get(tabId)!
+
+      expect(tab.id).to.equal(attachInfo.id)
+      expect(tab.windowId).to.equal(attachInfo.newWindowId)
+      expect(tab.domain).to.equal('slam.com')
```
Comment (will-weiss) [inline src/tests/integration/monitor-tabs.test.ts:93]: This is misleading as it appears to indicate that the tab domain came from those sent in the callback, which they didn't, let's remove these.

File: src/tests/integration/monitor-tabs.test.ts [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+    it('has no effect for active tabs of other windows', () => {
+      const { tabs } = mocks.getState()
+      expect(tabs.get(17)).to.have.property('active', true)
+    })
+  })
+
+  describe('chrome.tabs.onReplaced', () => {
+    it("replaces a tab (create and remove) and updates store's state", () => {
+      expect(mocks.chrome.tabs.onCreated.addListener).to.have.callCount(1)
```
Comment (will-weiss) [inline src/tests/integration/monitor-tabs.test.ts:117]: I don't understand, this never calls the callback from `chrome.tabs.onReplaced`

File: src/tests/integration/monitor-tabs.test.ts [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+  describe('chrome.windows.onRemoved', () => {
+    it("Deletes a removed window from store's state", () => {
+      const [callback] = mocks.chrome.windows.onRemoved.addListener.firstCall.args
+      const removedWindow = { id: 1, focused: false }
+      callback(removedWindow.id)
+
+      const state = mocks.getState()
+
+      if (state.mostRecentAction.type !== 'chrome.windows.onRemoved') {
```
Comment (will-weiss) [inline src/tests/integration/monitor-tabs.test.ts:178]: Similar note as above - we're trying to test the specific behavior of the responder, not the generic behavior that sets the `mostRecentAction`

File: src/tests/integration/monitor-tabs.test.ts [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+  })
+
+  describe('chrome.windows.onFocusChanged', () => {
+    it("Changes windows focus status in store's state", () => {
+      const [callback] = mocks.chrome.windows.onFocusChanged.addListener.firstCall.args
+      const changedWindow = { id: 3, focused: true }
+      callback(changedWindow.id)
+      const state = mocks.getState()
+
+      if (state.mostRecentAction.type !== 'chrome.windows.onFocusChanged') {
```
Comment (will-weiss) [inline src/tests/integration/monitor-tabs.test.ts:193]: Similar note as above - we're trying to test the specific behavior of the responder, not the generic behavior that sets the mostRecentAction

File: src/tests/integration/monitor-tabs.test.ts (hunk 3 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts:57]: Resolved. Thanks!

File: src/tests/integration/monitor-tabs.test.ts (hunk 4 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts:82]: also replaced tabId with attachInfo.id

File: src/tests/integration/monitor-tabs.test.ts (hunk 5 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts:86]: Also, replaced tabId with attachInfo.id. resolved.

File: src/tests/integration/monitor-tabs.test.ts (hunk 7 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts:112]:
```
it("replaces a tab (create and remove) and updates store's state", () => {
      expect(mocks.chrome.tabs.onCreated.addListener).to.have.callCount(1)
      const [callbackOnCreacted] = mocks.chrome.tabs.onCreated.addListener.firstCall.args

      const addedTabId = 11
      const addedInfo = {
        id: 11,
        windowId: 1,
        active: false,
        url: 'https://foo.com/abc',
      }

      // adding previouly deleted tab
      callbackOnCreacted(addedInfo)

      // tab 18 was created from previous test
      const removedTabId = 18
      const removeInfo = {
        id: 18,
        windowId: 1,
        active: false,
        url: 'https://new.com/abc',
      }

      // console.log(removeInfo.url)

      const [callbackOnReplaced] = mocks.chrome.tabs.onReplaced.addListener.firstCall.args
      callbackOnReplaced(addedTabId, removedTabId)

      const stateOnReplaced = mocks.getState()
      const tabsNotReplaced = stateOnReplaced.tabs

      expect(tabsNotReplaced.get(removedTabId)).to.equal(undefined)
      expect(tabsNotReplaced.get(addedTabId)).to.have.property('id', addedInfo.id)
      expect(tabsNotReplaced.get(addedTabId)).to.have.property('windowId', removeInfo.windowId)
      expect(tabsNotReplaced.get(addedTabId)).to.have.property('active', removeInfo.active)
      expect(tabsNotReplaced.get(addedTabId)).to.have.property('url', removeInfo.url)
    })
```

Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts:112]: the Chrome.tabs.onReplace() function seems only replace the tab id, the rest of tab info are the same as deleted tab?

File: src/tests/integration/monitor-tabs.test.ts (hunk 2 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts:35]: But the listener callback function asking for tabId and removeInfo, right? 
chrome.tabs.onRemoved.addListener((tabId, removeInfo) => dispatch['chrome.tabs.onRemoved']({ tabId, removeInfo }))

File: src/tests/integration/monitor-tabs.test.ts (hunk 4 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts]: resolved

File: src/tests/integration/monitor-tabs.test.ts (hunk 6 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts:89]: Refactor as discussed.

File: src/tests/integration/monitor-tabs.test.ts (hunk 7 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts:112]: Refactor as discussed.

File: src/tests/integration/monitor-tabs.test.ts (hunk 6 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts:89]: Remove assertion syntax for domain.

File: src/tests/integration/monitor-tabs.test.ts (hunk 2 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts:35]: Refactor as instructed. Test passed

File: src/tests/integration/monitor-tabs.test.ts (hunk 1 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts]: Refactor as instructed

File: src/tests/integration/monitor-tabs.test.ts (hunk 8 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts]: expect(state.focusedWindowId).not.eql(removedWindow.id)

File: src/tests/integration/monitor-tabs.test.ts (hunk 9 above)
Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts]: expect(state.focusedWindowId).to.equal(changedWindow.id)

File: src/tests/integration/monitor-tabs.test.ts [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+      expect(tabs.get(existingTab.id)).to.equal(undefined)
+      expect(tabs.get(newTabId)).to.have.property('id', newTabId)
+      expect(tabs.get(newTabId)).to.have.property('windowId', existingTab.windowId)
+      expect(tabs.get(newTabId)).to.have.property('active', existingTab.active)
+      expect(tabs.get(newTabId)).to.have.property('url', existingTab.url)
+    })
+  })
+
+  describe('chrome.windows.onCreated', () => {
+    it("Adds a newly created window to store's state", () => {
```
Comment (will-weiss) [inline src/tests/integration/monitor-tabs.test.ts:117]: Let's use lowercase as we do everywhere else

File: src/tests/integration/monitor-tabs.test.ts [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+      expect(tab.id).to.equal(changeInfo.id)
+      expect(tab.windowId).to.equal(changeInfo.windowId)
+      expect(tab.domain).to.equal('changeinfo.com')
+      expect(tab.url).to.equal('https://changeinfo.com/abc')
+    })
+  })
+
+  describe('chrome.tabs.onAttached', () => {
+    it("Attaches a tab to another window and updates the store's state", () => {
+      const [callbackOnAttached] = mocks.chrome.tabs.onAttached.addListener.firstCall.args
```
Comment (will-weiss) [inline src/tests/integration/monitor-tabs.test.ts:63]: Let's just call this `callback` as we do throughout the file

Comment (hankeliu2015) [reply src/tests/integration/monitor-tabs.test.ts:63]: mock.restore() 
Great question. Yes, neglecting to reset/restore your mocks can cause issues. We’re using sinon-chrome which is a library someone made with a bunch of stubs representing the chrome API. You’ll see that in the teardown function we call chrome.reset() . That teardown function gets called after all tests run in whatever describe block the createMocks function got called in.

https://github.com/morehumaninternet/roar-extension/blob/54bc39645caaef5e0992900f8ac9ee190b9083ef/src/tests/integration/mocks.tsx#L7
https://github.com/morehumaninternet/roar-extension/blob/54bc39645caaef5e0992900f8ac9ee190b9083ef/src/tests/integration/mocks.tsx#L103
https://github.com/morehumaninternet/roar-extension/blob/54bc39645caaef5e0992900f8ac9ee190b9083ef/src/tests/integration/mocks.tsx#L113
If locally you comment out the chrome.reset function you’ll see things go haywire very fast.

**Review by hankeliu2015 — APPROVED** [review]

**Pull request merged**